applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/mechanical_explicit_strategy.hpp

```diff
@@ -461,6 +486,103 @@ class MechanicalExplicitStrategy
     ///@name Private Operations
     ///@{
 
+    /**
+     * @brief This method computes the explicit contribution of the constraints
+     * @param pScheme The pointer to the integration scheme used
+     * @param rModelPart The model part which defines the problem
+     */
+    void ComputeExplicitConstraintConstribution(
+        typename TSchemeType::Pointer pScheme,
+        ModelPart& rModelPart
+        )
+    {
+        // The current process info
+        ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+
+        // Contributions to the system
+        DenseMatrixType transformation_matrix = DenseMatrixType(0, 0);
+        DenseVectorType constant_vector = DenseVectorType(0);
+
+        // Vector of dofs values
+        Vector slave_dofs_values, master_dofs_values;
+
+        // Vector containing the localization in the system of the different terms
+        typename MasterSlaveConstraint::DofPointerVectorType slave_equation_dofs, master_equation_dofs;
+
+        const int number_of_constraints = static_cast<int>(rModelPart.MasterSlaveConstraints().size());
+
+        // Setting to zero the slave dofs
+        #pragma omp parallel firstprivate(slave_equation_dofs)
+        {
+            #pragma omp for schedule(guided, 512)
+            for (int i_const = 0; i_const < number_of_constraints; ++i_const) {
+                auto it_const = rModelPart.MasterSlaveConstraints().begin() + i_const;
+
+                // Detect if the constraint is active or not. If the user did not make any choice the constraint
+                // It is active by default
+                bool constraint_is_active = true;
+                if (it_const->IsDefined(ACTIVE))
+                    constraint_is_active = it_const->Is(ACTIVE);
+
+                if (constraint_is_active) {
+                    slave_equation_dofs = it_const->GetSlaveDofsVector();
+
+                    for (IndexType i = 0; i < slave_equation_dofs.size(); ++i) {
+                        #pragma omp critical
```

Comment:
?critical?
this is a performance killer!!

Reply:
I know, the alternative is to construct a dof set in the Initialize with the slave dofs

Reply:
you still have a very severe bottleneck there. Consider doing something like

~~~cpp
double& value = slave_equation_dofs[i]->GetSolutionStepValue();
#pragma omp atomic
value *= 0.0;
~~~

Reply:
Nice trick I will try

Reply:
It works, I din't know *= was compatible with atomic. I thought was only -= and +=